### PR TITLE
ensure analog signal proxy and analog signal behave the same for time slices 

### PR DESCRIPTION
**Pull request by zm711**
In #1017 there is a discrepancy between 
`lazy=True` vs `lazy=False` because one uses `AnalogSignal` and one uses `AnalogSignalProxy`. This is likely due to the fact that `AnalogSignalProxy` is missing an `np.ceil` when calculating the `i_stop` which is included when calculating `i_start`. `AnalogSignal` accomplishes this with `np.rint` which when including the `astype(int64)` does the same as `int(np.ceil())`

https://github.com/NeuralEnsemble/python-neo/blob/53b202b2a6ae306392063d1dd2438323b4815de3/neo/io/proxyobjects.py#L191-L204

Let me know if you want a test added for this.

### CODE REVIEW

**Comment by zm711** [comment]
Actually the tests failed because currently the `AnalogSignalProxy` rounds down, but if the point of time slice is to include any information from within the time slice then even if we go slightly over then we need to round up to the next `i_index`. So in the tests our `t_stop = 5.00001` which is much less than a sample which for the test is `0.0001`. In the old version of the `AnalogSignalProxy` because 0.000001<0.0001 it rounds down and only includes signal information up to 5 which means we are losing one "index" of information. 

I would say based on #1017 this is not the logic users expect. If I want a time slice of 5.00001 seconds then I want the whole extra sample which would be 0.0001. 

But if the original behavior is desired then this can pull request can be closed and it needs to be made clear that the `AnalogSignalProxy` will only round down to the nearest sample rather than round up like `AnalogSignal`.

File: neo/test/iotest/test_proxyobjects.py [hunk 1]
```diff
@@ -54,8 +54,8 @@ def test_AnalogSignalProxy(self):
         # ceil next sample when slicing
         anasig = proxy_anasig.load(time_slice=(1.99999 * pq.s, 5.000001 * pq.s))
         assert anasig.t_start == 2. * pq.s
-        assert anasig.duration == 3. * pq.s
-        assert anasig.shape == (30000, 8)
+        assert anasig.duration == 3.0001 * pq.s
+        assert anasig.shape == (30001, 8)
```
Comment (samuelgarcia) [inline neo/test/iotest/test_proxyobjects.py:58]: I think theses test were design with np.round in mind.

**Comment by samuelgarcia** [comment]
Hi Zach,
thanks a lot for finding the little but very important detail!
I was not aware that astype(int64) was equivalent to int(np.ceil()). I was thinking it was rounding.
Do you have reference for that ?

**Comment by JuliaSprenger** [comment]
Hi! I am still a bit confused as `np.rint(a).astype('int64') == int(np.ceil(a))` does only hold for numbers with a decimal component `>=0.5`.  I agree that we should have the same behaviour for proxy and non-proxy objects. But why can't we use the same mechanism for both implementations?

**Comment by zm711** [comment]
Howdy Sam @samuelgarcia  & @JuliaSprenger 

Response to Sam:
Actually I was referring to the np.rint which would improve the behavior and less the casting with astype (although np.rint can also round down as opposed to np.ceil)—sorry for being unclear. The alternative would be change both int+np.ceil into np.rint. Which is what you’ve just said @JuliaSprenger 

Response to Julia:
Happy to switch them all to np.rint if that is the preferred choice. I think the np.ceil gives the user the sample they request even if it is a partial sample, but since neo has never guaranteed that it is fine to use the original behavior.

File: neo/test/iotest/test_proxyobjects.py (hunk 1 above)
Comment (zm711) [reply neo/test/iotest/test_proxyobjects.py]: Reverted the tests and switched to rint

**Comment by JuliaSprenger** [comment]
I would be happy with this solution. @samuelgarcia You too?

**Comment by samuelgarcia** [comment]
I reading in the comment `# the i_start is necessary ceil` I think I wrote this but do not remember exactly the why.

I think maybe one possible behavior I add in mind was : "ceil the t_start and round the duration" (and then compute i_stop). The actual proposal is round both t_start and t_stop which is not exactly the same.

For sure we need to harmonize between all part of neo : neo object and proxy object.

The actual AnalogSignal.time_slice have round t_start and then round the duration. Which is also another implementation!

I think rounding the duration is important. The question is more rounding t_start of ceiling t_start.

**Comment by samuelgarcia** [comment]
OK. Lets align `AnalogSignalProxy` with the historical `AnalogSignal.time_slice()`
Which is round(t_start) and round(duration).
I let you double check this behavior:
https://github.com/NeuralEnsemble/python-neo/blob/master/neo/core/analogsignal.py#L443
https://github.com/NeuralEnsemble/python-neo/blob/master/neo/core/analogsignal.py#L437

Thanks a lot for caching and patching this and sorry for so long latencies.
Detail are super importants!

**Comment by zm711** [comment]
@samuelgarcia ready for you to check. I kept the syntax the same as AnalogSignal to make it easier to compare in the future.

**Comment by samuelgarcia** [comment]
Merci  beaucoup.
This OK for me.
@JuliaSprenger do you want to have a look ?

**Review by JuliaSprenger — APPROVED** [review]

**Comment by JuliaSprenger** [comment]
Looks also good to me! Thanks for making the slicing more consistent!

**Comment by zm711** [comment]
@JuliaSprenger just in case you like cleaning up the issue tracker this as far as I can tell fixes #1017.  Thanks for all help!

**Pull request merged**